### ISSUE TITLE
Goals for next release?

### ISSUE THREAD

**Issue by Kiv** [leela-zero/leela-zero#1138]
I've been following the project and running autogtp on my 1080 Ti since December, and I'm interested in doing more to help out whether that's on the development side, or other areas like documentation and coordination.

Is there any info written down about what are the near term priorities for the project, and what work is available to be done? 

I'm struggling to read through the issue tracker as a lot of the open issues are just general discussion and not directly actionable, or issues that've been open for weeks/months and it's unclear if anyone is still working on them. So I apologize in advance if this has already been discussed recently.

**Comment by hydrogenpi** [comment]
In nutshell:

Recently SWA injection added to help with variety to hopefully boost the 10 block up some more plus the learning rate was decreased one last final time. So there won't be much left in the 10 block until we switch to 20 block. 

Problem is no one is getting decent results on the higher sized blocks anymore... and so not sure where to go from here.

**Comment by PhilipFRipper** [comment]
That graph displays a horrible understanding.  Also, GCP has a job.  He does this outside his job because he wants to.  I hope we can keep speculation and stuff off so they can work without harassment.   All of us look forward to updates but we have no entitlement to them, hydro.  TY.

**Comment by gcp** [comment]
>learning rate was decreased one last final time.

Learning rate was increased when SWA was introduced.

**Comment by gcp** [comment]
>Is there any info written down about what are the near term priorities for the project, and what work is available to be done?

The most important things in the engine right now are likely dealing with the memory usage (in a way that doesn't make the code harder to work with) and perhaps optimizing the OpenCL code, especially near the Winograd transforms some gain may be able to be made.

There's also a ton of ideas posted here where nobody wants to test things, so those generally don't end up going anywhere. The search can potentially be improved a ton, but this may not directly be usable for the self-play learning.

The engine itself is in a fairly good state right now, there's nothing particular to do when it just works. Supporting multiple board sizes from the same exe would be nice, but it's probably going to be fairly invasive work.

**Comment by gcp** [comment]
Multi-GPU support for the TensorFlow code would also likely be welcomed by a lot of people.

**Comment by bood** [comment]
> Multi-GPU support for the TensorFlow code would also likely be welcomed by a lot of people

AFAIK, @godmoves has made a working version but he's still testing it

**Comment by gcp** [comment]
I agree that would be interesting to have. We may be able to salvage the fp16 code from CLBlast upstream.

**Comment by isty2e** [comment]
Speaking of fp16, I remember that fp16 and mixed precision training did not work very well (as in #562). I wonder how bad it was; If the symptom was not severe, could there be any possibility that it could be saved by SWA, though I doubt it?

**Comment by Hersmunch** [comment]
I notice there is a new CLBlast release (https://github.com/CNugteren/CLBlast/releases) but I don't know whether any of the changes could help us:

> CLBlast version 1.3.0. Changes since previous release (version 1.2.0):
> 
> * Re-designed and integrated the auto-tuner, no more dependency on CLTune
> * Made it possible to override the tuning parameters in the clients straight from JSON tuning files
> * Added OpenCL pre-processor to unroll loops and perform array-to-register promotions for compilers
> which don't do this themselves (ARM Mali) - greatly improves performance on these platforms
> * Added first tuners for the TRSV (block size) and TRSM (invert kernel) routines
> * Added an optional argument to the GEMM routine to provide a pre-allocated temporary buffer
> * Fixed an issue with a crashing/hanging AMD APP compiler with the TRSM routine (invert kernel)
> * Improved compilation time by splitting the tuning database into multiple compilation units
> * Various minor fixes and enhancements
> * Added tuned parameters for various devices (see README)
> * Added the RetrieveParameters function to the API to be able to inspect the tuning parameters
> * Added a strided-batched (not part of the BLAS standard) routine, faster but less generic compared
> to the existing xGEMMBATCHED routines:
> SGEMMSTRIDEDBATCHED/DGEMMSTRIDEDBATCHED/CGEMMSTRIDEDBATCHED/ZGEMMSTRIDEDBATCHED/HGEMMSTRIDEDBATCHED

**Comment by hydrogenpi** [comment]
Seems like there is a new kid on the block:

https://github.com/unrealgo/unrealgo/issues/1

**Comment by gcp** [comment]
>Added OpenCL pre-processor to unroll loops and perform array-to-register promotions for compilers
which don't do this themselves (ARM Mali) - greatly improves performance on these platforms

This is potentially interesting for people who want to try Android ports.

The others don't look particularly interesting, unless the new auto-tuner is smarter.

**Comment by gcp** [comment]
I think there are some important fixes for a new release, but I want to add some time control improvements still. (Move slower, and use all time in Japanese byo yomi).

**Comment by Ttl** [comment]
CLBlast recently added a new SGEMM kernel which should be faster especially on Intel GPUs. On my laptop with i5-8250U CPU, I measured 115 GFLOPS with the old kernel and 165 GFLOPS with the new one using CLBlast tuner. Unfortunately the new kernel needs the first matrix transposed and returns the output also transposed. I tried to make it work with the current code, but it looks like too much trouble for the performance gain. If someone really cares about performance on Intel GPUs it might be worth investigating.
